Implementation Of Conference Information View

at-robins opened the pull request:
A new view displaying conference information has been defined in the template info.scala.html.
The conference information stored on the server is Markdown formatted therefore the commonmark library is used in the server side conference model to convert the data to HTML.
All tests perform without error, however a encoding conflict inside the database seems to cause erroneous display of some special characters such as ä, ü, ö, etc..

at-robins committed:
Implemented Conference Information View

A new view displaying conference information has been defined in the template info.scala.html.
The conference information stored on the server is Markdown formatted therefore the commonmark library is used in the server side conference model to convert the data to HTML.
All tests perform without error, however a encoding conflict inside the database seems to cause erroneous display of some special characters such as ä, ü, ö, etc..

(cherry picked from commit 2ad1e6c)

cgars commented:
## Some general comments
* Well done!
* Rebase again, and push (maybe with force) this PR instead of creating a new one.
* Only show the Information stuff if there is data.
* Maybe we could have the info view not as an additional tab but below the conference (extended) view

cgars commented on app/controllers/Application.scala:
```diff
@@ -43,6 +43,17 @@ class Application(implicit val env: Environment[Login, CachedCookieAuthenticator
+  /**
+    * Handles the display of the information tab for a specific conference.
+    *
+    * @param confId - the conference ID
```
For the sake of uniformity, I would probably go without doc here.

cgars commented on app/models/Conference.scala:
```diff
@@ -123,6 +130,21 @@ class Conference extends Model with Owned with Tagged {
     this.mtime = new DateTime(DateTimeZone.UTC)
   }
+
+  /**
```
same here

cgars commented on app/models/Conference.scala:
```diff
@@ -123,6 +130,21 @@ class Conference extends Model with Owned with Tagged {
+     * Parse the Markdown formatted conference information and render it to HTML.
+     */
+    // TODO: write test case
+    return HtmlRenderer.builder().build().render(Parser.builder().build().parse(this.info))
```
We could consider not instantiating this all the time. Good enough for the time being though.

at-robins committed:
Integration Of Information View Into Conference View

The conference information view has been removed as self-contained view. Instead it has been integrated into the conference view as additional panel. The design/layout may need further improvements.
Additionally unnecessary comments and documentation have been removed.
Furthermore the Markdown parser has been optimised to utilise a static parser and HTML renderer instead of recreating them for every conference.
The encoding problem is still unresolved.

at-robins commented:
The conference information view is currently implemented as second box below the conference box. Maybe I should change this to be the same as the "Administration" function when logged in? I guess it comes down to personal taste.

cgars reviewed (changes requested):
Way better now. 
I have an issue with the reuse of the info view. We have the conference uuid in there twice now. Also, the "no info here" warning is kind of obsolete

at-robins committed:
Complete Removal Of Conference Information View

The conference information view, previously reused as template, has been removed completely and fully integrated into the conference view.

at-robins commented:
I removed the view completely now.
Should I write a test case for the Markdown-parsing function? It is pretty much just a few library calls, so I do not think it to be really necessary.

cgars commented on app/views/conference.scala.html:
```diff
@@ -77,4 +77,16 @@ <h3>Administration</h3>
     </div>
+
+    <!-- The conference info panel -->
+    @if(conference.info != null) {
```
Have you actually checked whether this is working? I kind of doubt that it will always be null if empty. You could start by testing for such cases

at-robins committed:
Fixed Conference Information Display Condition

An additional condition has been added to the conference view to ensure empty information is not displayed.

at-robins committed:
Markdown Parsing As Static Function

The Markdown parsing function of the conference model has been made static to enhance reusability and simplify testing.
Additionally an unused function for a planned replacement of the current conference description format (HTML) with Markdown has been implemented.

at-robins committed:
Test Cases For The Conference Markdown Parsing Added

Test cases for parsing markdown formatted conference information and converting it to HTML code have been implemented. The tests cover most of the basic applications of Markdown and its specification.
The used parser library performs well in all test cases except removing some whitespace at line breaks, which seems to be a minor issue, if any at all.

cgars commented on test/models/ConferenceTest.scala:
```diff
@@ -0,0 +1,151 @@
+     * Mostly a placeholder for potential future changes, which need to be tested differently.
+     */
+    val r : Random = new Random()
+    var testingInfo : String = ""
```
Testing this (empty string) would actually be an interesting corner case

at-robins replied:
Actually empty strings are tested implicitly at two points. First in 
```
  def testConvertMarkdownToHTML() : Unit = {
    val r : Random = new Random()
    // test nulls throwing a NullPointerException
    intercept[NullPointerException] {
      Conference.convertMarkdownToHTML(null)
    }
    // test empty markdown returning empty html
    assert(Conference.convertMarkdownToHTML("") == "")
```
, which is currently equivalent to the instance functions.
And a second time as part of the random testing since strings of length zero are allowed.
I, however, added the empty string corner case explicitly to the instance function tests for the sake of completeness.

cgars commented on test/models/ConferenceTest.scala:
```diff
@@ -0,0 +1,151 @@
+        None, None, Some(false), Some(true), Some(false), Some(true), None, None, None, info = Some(testingInfo))
+      assert(testingConf.getInfoAsHTML() == Conference.convertMarkdownToHTML(testingInfo))
+    }
+    // test nulls throwing a NullPointerException
```
But should they. I know they do however should they really.  Are you not, you trying to avoid it?

at-robins replied:
Sure, I guess you mean the XSS protection is still missing?

cgars replied:
Let me ask differently. where would the np happen and how could it be avoided?

at-robins committed:
Added Empty Strings To Markdown Test Cases

Empty strings are now explicitly tested for the Markdown-to-HTML-converting functions of the conference model.

at-robins committed:
Fixed Coding Style

Some minor coding style issues in the conference model.

at-robins committed:
Omitted NullPointerException In Markdown Parsing Functions

The conversion of conference information and description from Markdown into HTML no longer throws a NullPointerExpection upon those fields being null. Instead an empty string is returned.
The static parsing function however still throws an exception.
